gh-142418: Fix inspect.iscoroutinefunction for marked functools.partial and partialmethod objects

Joshua-Ward1 opened the pull request:
Fixes gh-142418.  

This PR updates `inspect.iscoroutinefunction()` so that coroutine markers
applied via `inspect.markcoroutinefunction()` are correctly detected on
wrapped callables, including `functools.partial` and
`functools.partialmethod` objects.  

Previously, `iscoroutinefunction()` only checked the marker after unwrapping
the callable, which caused false negatives for marked `partial` and
`partialmethod` objects. The new logic checks for the marker at each unwrap
stage, with cycle protection, ensuring that any explicitly marked wrapper is
recognized as a coroutine function.  

This change also adds regression tests verifying correct behavior for marked
and unmarked `functools.partial` and `functools.partialmethod` objects, and
includes a NEWS entry documenting the fix.


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--142503.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

Joshua-Ward1 committed:
Doc: clarify -i interaction with PYTHONSTARTUP

Joshua-Ward1 committed:
gh-142418: Fix inspect.iscoroutinefunction for marked functools.partial and partialmethod objects

x42005e1f commented on Lib/inspect.py:
```diff
@@ -306,10 +306,33 @@ def isgeneratorfunction(obj):
+    while True:
+        if id(f) in visited:
+            return False
+        visited.add(id(f))
```
Partial/method objects cannot create reference cycles (or at least should not), so I do not think that handling this case is worth the performance degradation of the function call. Can you imagine calling such an object?

Joshua-Ward1 committed:
Fix NEWS entry formatting for gh-142418

Corrected the NEWS entry

Joshua-Ward1 committed:
Rename NEWS entry to follow format

Updated to follow format

Joshua-Ward1 committed:
Fixed Filename

x42005e1f commented on Lib/inspect.py:
```diff
@@ -306,10 +306,33 @@ def isgeneratorfunction(obj):
+            f = getattr(f, 'func')
+            continue
+
+        if ismethod(f):
+            f = f.__func__
+            continue
```
Method objects cannot be marked, so it makes sense to move this check to the beginning for optimization.

x42005e1f commented on Lib/inspect.py:
```diff
@@ -306,10 +306,33 @@ def isgeneratorfunction(obj):
+
+        if isinstance(f, functools.partialmethod):
+            f = getattr(f, 'func')
+            continue
```
If you want to handle `partialmethod` objects as in the current `iscoroutinefunction()` behavior, it would be more efficient to do so in the existing check for `partial` objects:

```python
if isinstance(f, (functools.partial, functools.partialmethod)):
    f = f.func
    continue
```

Note that using `getattr()` here makes no sense.

Joshua-Ward1 committed:
gh-142418: Remove unnecessary cycle detection from _has_coroutine_mark

Partial and method objects cannot form reference cycles in this context, so
the cycle-detection logic in _has_coroutine_mark introduced unnecessary
overhead. This change removes the visited-set check and keeps the function
as a simple unwrapping loop while still correctly detecting explicitly
marked coroutine wrappers.

Joshua-Ward1 committed:
gh-142418: Simplify _has_coroutine_mark by removing unnecessary cycle-detection logic

x42005e1f commented on Lib/inspect.py:
```diff
@@ -306,10 +306,22 @@ def isgeneratorfunction(obj):
+            f = f.__func__
+            continue
+
+        # partial and partialmethod share .func
+        if isinstance(f, (functools.partial, functools.partialmethod)):
+            f = f.func
+            continue
```
Why did you remove the check for regular functions created by `partialmethod`?

Joshua-Ward1 replied:
Whoops, fixed.

Joshua-Ward1 committed:
gh-142418: Restore __partialmethod__ handling in _has_coroutine_mark

x42005e1f commented on Lib/inspect.py:
```diff
@@ -306,10 +306,29 @@ def isgeneratorfunction(obj):
+    while True:
+        # Direct marker check
+        if getattr(f, "_is_coroutine_marker", None) is _is_coroutine_mark:
+            return True
```
It makes sense to do this after checking for method objects, since they are not marked anyway.

Joshua-Ward1 committed:
Reorder coroutine marker check after method unwrapping

x42005e1f commented on Lib/inspect.py:
```diff
@@ -306,10 +306,28 @@ def isgeneratorfunction(obj):
+        if getattr(f, "_is_coroutine_marker", None) is _is_coroutine_mark:
+            return True
+
+        # Functions created by partialmethod descriptors keep a __partialmethod__ reference
+        pm = getattr(f, "__partialmethod__", None)
+        if isinstance(pm, functools.partialmethod):
+            f = pm
+            continue
```
Well, this can also be moved forward by one block to avoid the time spent on obtaining the attribute when it is not necessary. I hope I have not bored you with these micro-optimizations.

x42005e1f commented on Lib/inspect.py:
```diff
@@ -306,10 +306,28 @@ def isgeneratorfunction(obj):
+            continue
+
+        # partial and partialmethod share .func
+        if isinstance(f, (functools.partial, functools.partialmethod)):
```
I will also add, "for the record", why I do not handle `partialmethod` objects in this way in the code attached to the original issue. They are not callable, and therefore applying `markcoroutinefunction()` to them is incorrect, which means they should not be checked. Being defined as a class member, accessing the corresponding attribute will return a regular function object created by `partialmethod` (or a method object for such a function, if via an instance). Therefore, there is no point in unnecessary iteration, and you can go straight to `pm.func` (see the block above).

kumaraditya303 commented on Doc/using/cmdline.rst:
```diff
@@ -302,16 +302,22 @@ Miscellaneous options
 
-   Interactive mode will start even when :data:`sys.stdin` does not appear to be a terminal. The
-   :envvar:`PYTHONSTARTUP` file is not read.
+   In these "execute then interact" cases, Python runs the script or command
```
This seems unrelated changes

picnixz commented on Lib/inspect.py:
```diff
@@ -306,10 +306,28 @@ def isgeneratorfunction(obj):
-        f = f.__func__
-    f = functools._unwrap_partial(f)
-    return getattr(f, "_is_coroutine_marker", None) is _is_coroutine_mark
+    while True:
```
Please:

- Wrap all lines under 80 characters.
- Remove "obvious" comments. "Methods: unwrap first" is clear from the way you're doing it.
- Avoid blank lines. The standard library usually tries to avoid expanding the code vertically.

picnixz commented on Lib/inspect.py:
```diff
@@ -306,10 +306,28 @@ def isgeneratorfunction(obj):
-    return getattr(f, "_is_coroutine_marker", None) is _is_coroutine_mark
+    while True:
+        # Methods: unwrap first (methods cannot be coroutine-marked)
+        if ismethod(f):
```
Why not doing `while ismethod(f): f = f.__func__` here?

x42005e1f replied:
```python
>>> from functools import partialmethod
>>> class MyClass:
...     def a(self): ...
...     b = partialmethod(a)
>>> obj = MyClass()
>>> obj.b
functools.partial(<bound method MyClass.a of <__main__.MyClass object at 0x7fd343f4cc20>>)
>>> obj.b.func.__func__
<function MyClass.a at 0x7fd343f42cf0>
```

When unwrapping the reference to `obj.b`, the `functools.partial` object comes before the method object, so the assumption that method objects always precede other objects is incorrect.

```python
>>> from functools import partialmethod
>>> class MyFirstClass:
...     def f(self, other): ...
>>> first = MyFirstClass()
>>> class MySecondClass:
...     g = partialmethod(first.f)
>>> second = MySecondClass()
>>> second.g
<bound method partialmethod._make_unbound_method.<locals>._method of <__main__.MySecondClass object at 0x7fd343f4d400>>
>>> second.g.__func__.__partialmethod__.func
<bound method MyFirstClass.f of <__main__.MyFirstClass object at 0x7fd343f4d2b0>>
```

Well, it seems to me that the current implementation (in the main branch) is not very well thought out. Personally, I would not rely on it.

x42005e1f replied:
If your comment is only about why there is `if` instead of `while`... why do we need a nested loop when we already have an outer one? Moreover, how often are methods that refer to other methods used? Does it even make sense to try to optimize this case? I think it is sufficient and simpler to rely on the outer loop.

As you can easily see, the code in this PR is very similar to the one I attached to the original issue, except for the order of the blocks and some points borrowed from the original code (AI-generated or just copy-paste?). I use something similar in [the implementation of similar functions in my library](https://github.com/x42005e1f/aiologic/blob/baf8187d391f16e9b2f75044eaf9b95cd7da7f5b/src/aiologic/meta/_inspect.py#L445-L485), although they are more general in nature.

picnixz commented on Lib/inspect.py:
```diff
@@ -306,10 +306,28 @@ def isgeneratorfunction(obj):
+        if getattr(f, "_is_coroutine_marker", None) is _is_coroutine_mark:
+            return True
+
+        # Functions created by partialmethod descriptors keep a __partialmethod__ reference
+        pm = getattr(f, "__partialmethod__", None)
+        if isinstance(pm, functools.partialmethod):
+            f = pm
+            continue
```
I think you should use `functools._unwrap_partialmethod` which handles both partial methods and partial functions (first it unwraps partial methods then unwraps partial functions)

x42005e1f replied:
Please read the discussion at #142505.
